### PR TITLE
pkg/prometheus: add SendExemplars in RemoteWrite

### PR DESCRIPTION
**Pull request by coryrjohnson**
Adds send_exemplars field to the remote_write configuration in Prometheus

### CODE REVIEW

File: pkg/apis/monitoring/v1/types.go [hunk 1]
```diff
@@ -652,6 +652,8 @@ type RemoteWriteSpec struct {
 	// name is used in metrics and logging in order to differentiate queues.
 	// Only valid in Prometheus versions 2.15.0 and newer.
 	Name string `json:"name,omitempty"`
+	// Only valid in Prometheus versions 2.15.0 and newer.
```
Comment (metalmatze) [inline pkg/apis/monitoring/v1/types.go:655]:
```suggestion
	// Only valid in Prometheus versions 2.27.0 and newer.
```

File: example/prometheus-operator-crd/monitoring.coreos.com_prometheuses.yaml [hunk 2]
```diff
@@ -4154,6 +4154,9 @@ spec:
                     remoteTimeout:
                       description: Timeout for requests to the remote write endpoint.
                       type: string
+                    sendExemplars:
+                      description: Only valid in Prometheus versions 2.15.0 and newer.
```
Comment (metalmatze) [inline example/prometheus-operator-crd/monitoring.coreos.com_prometheuses.yaml:4158]:
```suggestion
                      description: Only valid in Prometheus versions 2.27.0 and newer.
```

File: bundle.yaml [hunk 3]
```diff
@@ -12383,6 +12383,9 @@ spec:
                     remoteTimeout:
                       description: Timeout for requests to the remote write endpoint.
                       type: string
+                    sendExemplars:
+                      description: Only valid in Prometheus versions 2.15.0 and newer.
```
Comment (metalmatze) [inline bundle.yaml:12387]:
```suggestion
                      description: Only valid in Prometheus versions 2.27.0 and newer.
```

**Review by metalmatze — CHANGES_REQUESTED** [review]
The code seems correct in terms of the semver however the comment is probably a copy-paste-mistake. Otherwise LGTM :+1:

**Comment by coryrjohnson** [comment]
> The code seems correct in terms of the semver however the comment is probably a copy-paste-mistake. Otherwise LGTM 👍

Thanks, updated.

File: Documentation/api.md [hunk 4]
```diff
@@ -745,6 +745,7 @@ RemoteWriteSpec defines the remote_write configuration for prometheus.
 | ----- | ----------- | ------ | -------- |
 | url | The URL of the endpoint to send samples to. | string | true |
 | name | The name of the remote write queue, must be unique if specified. The name is used in metrics and logging in order to differentiate queues. Only valid in Prometheus versions 2.15.0 and newer. | string | false |
+| sendExemplars | Enables sending of exemplars over remote write. Note that exemplar storage itself must be enabled for exemplars to be scraped in the first place.  Only valid in Prometheus versions 2.27.0 and newer. | boolean | false |
```
Comment (simonpasquier) [inline Documentation/api.md:748]: This file is generated from `pkg/apis/monitoring/v1/types.go` so the comment needs to go there.

Comment (simonpasquier) [reply Documentation/api.md:748]: Maybe also reference the `enableFeatures` field.

**Comment by coryrjohnson** [comment]
Let me know how that looks @simonpasquier

**Review by simonpasquier — COMMENTED** [review]
lgtm

File: pkg/apis/monitoring/v1/types.go [hunk 5]
```diff
@@ -653,6 +653,8 @@ type RemoteWriteSpec struct {
 	// name is used in metrics and logging in order to differentiate queues.
 	// Only valid in Prometheus versions 2.15.0 and newer.
 	Name string `json:"name,omitempty"`
+	// Enables sending of exemplars over remote write. Note that exemplar-storage itself must be enabled using the enableFeature option for exemplars to be scraped in the first place.  Only valid in Prometheus versions 2.27.0 and newer.
+	SendExemplars bool `json:"sendExemplars,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/types.go:657]: Can you make it *bool? This way, a nil value would stick to whatever default value is chosen by Prometheus.

```suggestion
	SendExemplars *bool `json:"sendExemplars,omitempty"`
```

File: pkg/prometheus/promcfg.go [hunk 6]
```diff
@@ -1483,6 +1483,10 @@ func (cg *ConfigGenerator) generateRemoteWriteConfig(
 			cfg = append(cfg, yaml.MapItem{Key: "name", Value: spec.Name})
 		}
 
+		if spec.SendExemplars && version.GTE(semver.MustParse("2.27.0")) {
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:1486]:
```suggestion
		if spec.SendExemplars != nil && version.GTE(semver.MustParse("2.27.0")) {
```

**Review by paulfantom — APPROVED** [review]
lgtm

**Review by simonpasquier — APPROVED** [review]

**Comment by simonpasquier** [comment]
@coryrjohnson can you run `make generate` and commit the changes?

**Comment by mvkrishna86** [comment]
@coryrjohnson, Thanks for the changes. Eagerly waiting for this to be merged to master.

**Pull request closed without merging**